bayem/vba.py

```diff
@@ -0,0 +1,519 @@
+            c0i, s0i = self.noise0[i].shape, self.noise0[i].scale
+            si, ci = self.s[i], self.c[i]
+            N = len(k[i])
+
+            # From the update equation
+            f_new += -si * ci / s0i + (N / 2 + c0i - 1) * (
+                np.log(si) + special.digamma(ci)
+            )
+            f_new += (
+                -0.5 * si * ci * (k[i].T @ k[i] + np.trace(self.L_inv @ J[i].T @ J[i]))
+            )
+            f_new += ci * np.log(si) + special.gammaln(ci)
+            f_new += ci - (ci - 1) * (np.log(si) + special.digamma(ci))
+            # constant terms to fix the evidence
+            f_new += (
+                -N / 2 * np.log(2 * np.pi) - special.gammaln(c0i) - c0i * np.log(s0i)
+            )
+
+        return f_new
+
+    def stop_criteria(self, f_new, i_iter):
+        self.n_trials += 1
+
+        if f_new > self.f_old:
+            self.n_trials = 0
+
+        # Update free energy here such that the "stop_criteria" is testable
+        # individually:
+        self.result.f_max = max(self.result.f_max, f_new)
+
+        # stop?
+        if self.n_trials >= self.options.maxtrials:
+            self.result.message = "Stopping because free energy did not "
+            self.result.message += (
+                f"increase within {self.options.maxtrials} iterations."
+            )
+            return True
+
+        if i_iter >= self.options.maxiter:
+            self.result.message = "Stopping because the maximum number of "
+            self.result.message = "iterations is reached."
+            return True
+
+        if abs(f_new - self.f_old) <= self.options.tolerance:
+            self.result.message = "Tolerance reached!"
+            self.result.success = True
+            return True
+
+        # go on.
+        self.f_old = f_new
+        return False
+
+
+
+
+class VBResult:
+    """
+    Somehow inspired by
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html
+    """
+
+    def __init__(self):
+        self.param = None
+        self.noise = {}
+        self.success = False
+        self.message = ""
+        self.free_energies = []
+        self.means = []
+        self.sds = []
+        self.shapes = []
+        self.scales = []
+        self.f_max = -np.inf
+        self.nit = 0
+        self.t = None
+
+    def __str__(self):
+        s = "### VB Result ###\n"
+
+        for member in [
+            "success",
+            "message",
+            "param",
+            "noise",
+            "free_energies",
+            "f_max",
+            "nit",
+            "t",
+        ]:
+            to_print = self.__dict__[member]
+            v = str(to_print).replace("\n", "\n" + " " * 17)
+            s += f"  {member+':':14s} {v}\n"
+
+        return s
+
+    def try_update(self, f_new, mean, precision, shapes, scales, names):
+        self.free_energies.append(f_new)
+        self.means.append(mean)
+        self.sds.append(MVN(mean, precision).std_diag)
```

Comment:
Why do you store only the diagonal entries. I would say that this should rather be a user choice. The standard case could actually be to store the complete precision (that's a little more data, but you don't have to invert the matrix).

Reply:
Add `store_full_precision=True` option.